library/std/src/io/buffered/bufwriter.rs

```diff
@@ -291,55 +291,94 @@ impl<W: Write> BufWriter<W> {
+            // we know that this will succeed in totality
+            self.buf.extend_from_slice(&buf[amt_buffered..]);
         }
+        Ok(())
     }
 
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
-        let total_len = bufs.iter().map(|b| b.len()).sum::<usize>();
-        if self.buf.len() + total_len > self.buf.capacity() {
-            self.flush_buf()?;
-        }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
-        if total_len >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write_vectored(bufs);
-            self.panicked = false;
-            r
+        if self.get_ref().is_write_vectored() {
+            let total_len: usize = bufs.iter().map(|buf| buf.len()).sum();
+
+            if total_len + self.buffer().len() > self.capacity() {
+                self.flush_buf()?;
+            }
+            if total_len >= self.buf.capacity() {
+                self.get_mut().write_vectored(bufs)
+            } else {
+                // Correctness note: we've already verified that none of these
+                // will overflow the buffer, because total_len < capacity
+                bufs.iter().for_each(|buf| self.buf.extend_from_slice(buf));
+                Ok(total_len)
+            }
         } else {
-            bufs.iter().for_each(|b| self.buf.extend_from_slice(b));
-            Ok(total_len)
+            // Because the inner writer doesn't have native vectored write
+            // support, we should take care of buffering together the individual
+            // incoming bufs, even if the *total* length is larger than our
+            // buffer. We only want to skip our buffer if an *individual* write
+            // exceeds our buffer capacity.
+            let mut total_buffered = 0;
+
+            for buf in bufs {
+                if total_buffered == 0 {
+                    if buf.len() + self.buffer().len() > self.capacity() {
+                        // If an individual write would overflow our remaining
+                        // capacity and we haven't buffered anything yet,
+                        // pre-flush before buffering (same as with regular
+                        // write())
+                        self.flush_buf()?;
+                    }
+
+                    if buf.len() > self.capacity() {
+                        // If an individual buffer exceeds our *total* capacity
+                        // and we haven't buffered anything yet, just forward
+                        // it to the underlying device
+                        return self.get_mut().write(buf);
+                    }
+                }
+
+                // Buffer as much as possible until we reach full capacity
+                total_buffered += self.write_to_buf(buf);
+                if self.buffer().len() == self.capacity() {
+                    break;
+                }
```

Comment:
If the total size is more than the capacity, this just fills up the capacity and then returns without looking at the remaining `IoSlice`s. Would it be better to `write()` at that point and continue? I assume that's not possible, because there's no way to return *'wrote N bytes, but also here's an error'*? Might be good to add a comment explaining that.

Reply:
Done!